src/main/java/bio/terra/app/controller/SearchApiController.java

```diff
@@ -94,7 +99,25 @@ private AuthenticatedUserRequest getAuthenticatedInfo() {
         @Valid @RequestParam(value = "offset", required = false, defaultValue = "0") Integer offset,
         @Valid @RequestParam(value = "limit", required = false, defaultValue = "1000") Integer limit
     ) {
-        SearchQueryResultModel searchQueryResultModel = new SearchQueryResultModel();
-        return new ResponseEntity<>(searchQueryResultModel, HttpStatus.OK);
+        List<String> accessibleIds =
+                iamService.listAuthorizedResources(getAuthenticatedInfo(), IamResourceType.DATASNAPSHOT)
+                        .stream()
+                        .map(UUID::toString)
+                        .collect(Collectors.toList());
+
+        List<String> requestIds = searchQueryRequest.getSnapshotIds();
+
+        Set<String> inAccessibleIds = new HashSet<>(requestIds);
+        inAccessibleIds.removeAll(accessibleIds);
+        if (!inAccessibleIds.isEmpty()) {
+                throw new IamForbiddenException("User '" + getAuthenticatedInfo().getEmail()
+                        + "' does not have required action: " + IamAction.READ_DATA
+                        + " on snapshot ids" + inAccessibleIds.toString());
```

Comment:
Unnecessary `toString()` (any object in a `String` `+` expression will have `.toString()` called on it automatically).
```suggestion
                        + " on snapshot ids" + inAccessibleIds);
```